### PR TITLE
Add support for cSHAKE to pycryptodome

### PR DESCRIPTION
**Pull request by msfschaffner**
This adds support for the customizable version of SHAKE (i.e., cSHAKE) to the `pycryptodome` package (see https://github.com/lowRISC/opentitan/issues/6694 for context).

The implementation piggybacks on the SHAKE implementations. 
The main two changes are the following:
1. Extend the constructor such that the *function* and *customization* strings can be specified when creating the hash object
2. Depending on the mode (plain SHAKE or cSHAKE), select the correct bit padding constant (`0x04` vs `0x1F`).

The unit tests are extended to also include a few patterns for cSHAKE on top of the ones already available for SHAKE.

I've also tried to reduce code duplication a bit by deriving all implementation variants (SHAKE*, cSHAKE*) from one base class (cSHAKE128).

See also the following two NIST pubs for reference:
- https://nvlpubs.nist.gov/nistpubs/FIPS/NIST.FIPS.202.pdf
- https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-185.pdf
- https://csrc.nist.gov/CSRC/media/Projects/Cryptographic-Standards-and-Guidelines/documents/examples/cSHAKE_samples.pdf

-------------------------------------

I've not yet PR'ed this to the upstream repo such that we can discuss this among ourselves first and iron out any issues we may encounter.

I intend to keep these commits separate also for the PR upstream to facilitate review, and then before the final merge I'll squash them all.

### CODE REVIEW

File: lib/Crypto/Hash/cSHAKE128.py [hunk 1]
```diff
@@ -28,39 +28,100 @@
...
     Do not instantiate directly.
     Use the :func:`new` function.
-
-    :ivar oid: ASN.1 Object ID
-    :vartype oid: string
     """
 
-    # ASN.1 Object ID
-    oid = "2.16.840.1.101.3.4.2.11"
+    # Parameters
```
Comment (msfschaffner) [inline lib/Crypto/Hash/cSHAKE128.py:94]: no OID is available for this algo variant.

File: lib/Crypto/Hash/cSHAKE128.py [hunk 2]
```diff
@@ -28,39 +28,100 @@
...
+
+        self.function = function
+        self.custom = custom
+
+        prefix = None
+        pad = 0x1F  # default to SHAKE
+        if function or custom:
+            prefix = encode_str(function) + encode_str(custom)
+            prefix = byte_align(prefix, self.prefix_alignment)
+            pad = 0x04  # for cSHAKE
```
Comment (msfschaffner) [inline lib/Crypto/Hash/cSHAKE128.py:110]: The the main functional changes are basically contained to these few lines, including lines 122 and 123 further below.

File: test_vectors/pycryptodome_test_vectors/Hash/SHA3/ShortMsgSamples_cSHAKE128.txt [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+MD = C1C36925B6409A04F1B504FCBCA9D82B4017277CB5ED2B2065FC1D3814D5AAF5
+
+# Sample 2
+NLen = 0
+SLen = 120
+Len = 1600
+N = 00
+S = 456D61696C205369676E6174757265
+Msg = 000102030405060708090A0B0C0D0E0F101112131415161718191A1B1C1D1E1F202122232425262728292A2B2C2D2E2F303132333435363738393A3B3C3D3E3F404142434445464748494A4B4C4D4E4F505152535455565758595A5B5C5D5E5F606162636465666768696A6B6C6D6E6F707172737475767778797A7B7C7D7E7F808182838485868788898A8B8C8D8E8F909192939495969798999A9B9C9D9E9FA0A1A2A3A4A5A6A7A8A9AAABACADAEAFB0B1B2B3B4B5B6B7B8B9BABBBCBDBEBFC0C1C2C3C4C5C6C7
+MD = C5221D50E4F822D96A2E8881A961420F294B7B24FE3D2094BAED2C6524CC166B
```
Comment (msfschaffner) [inline test_vectors/pycryptodome_test_vectors/Hash/SHA3/ShortMsgSamples_cSHAKE128.txt:19]: some more vectors for these two variants would be nice. 
unfortunately I haven't come across any.
maybe we can dump some from our SV UVM testbench?

**Comment by msfschaffner** [comment]
@imphil @rswarbrick @eunchan @udinator @cfrantz (I unfortunately can't add you as reviewers directly on this fork)

File: test_vectors/pycryptodome_test_vectors/Hash/SHA3/ShortMsgSamples_cSHAKE128.txt (hunk 3 above)
Comment (udinator) [reply test_vectors/pycryptodome_test_vectors/Hash/SHA3/ShortMsgSamples_cSHAKE128.txt:19]: I have a couple of test vector samples in `hw/dv/sv/test_vectors/vectors/xof/` from NIST, you can probably use those?

File: Doc/src/hash/shake256.rst [hunk 4]
```diff
@@ -7,7 +7,7 @@ As a XOF, SHAKE256 is a generalization of a cryptographic hash function.
 Instead of creating a fixed-length digest (e.g. 32 bytes like SHA-2/256),
 it can produce outputs of any desired length.
 
-Output bits do **not** depend of the output length.
+Output bits do **not** depend on the output length.
```
Comment (eunchan) [inline Doc/src/hash/shake256.rst:10]: Does not need this in this PR IMO.

**Review by eunchan — COMMENTED** [review]
Overall Looks good to me. :) Nice!

File: test_vectors/pycryptodome_test_vectors/Hash/SHA3/ShortMsgSamples_cSHAKE128.txt (hunk 3 above)
Comment (msfschaffner) [reply test_vectors/pycryptodome_test_vectors/Hash/SHA3/ShortMsgSamples_cSHAKE128.txt:19]: Are they by change the same ones from here?
https://csrc.nist.gov/CSRC/media/Projects/Cryptographic-Standards-and-Guidelines/documents/examples/cSHAKE_samples.pdf

These are the ones I have used here ;)...

Comment (udinator) [reply test_vectors/pycryptodome_test_vectors/Hash/SHA3/ShortMsgSamples_cSHAKE128.txt:19]: yup, unfortunately there aren't too many golden CSHAKE vectors...

**Review by udinator — APPROVED** [review]
lgtm!

**Comment by msfschaffner** [comment]
@imphil @rswarbrick 
OK if I move forward with this and switch our OT Python requirements to this lowRISC hosted fork until the patch has been upstreamed?

**Comment by imphil** [comment]
> OK if I move forward with this and switch our OT Python requirements to this lowRISC hosted fork until the patch has been upstreamed?

Sure, go ahead!

**Pull request merged**